src/System.Private.CoreLib/shared/System/Number.Grisu3.cs

```diff
@@ -311,410 +308,621 @@ internal static unsafe class Grisu3
 
-                        goto case 23;
-                    }
+                // The DiyFp.Multiply procedure rounds its result and tenMk is approximated too.
+                // The variable scaledW (as well as scaledBoundaryMinus/Plus) are now off by a small amount.
+                //
+                // In fact, scaledW - (w * 10^k) < 1ulp (unit in last place) of scaledW.
+                // In other words, let f = scaledW.f and e = scaledW.e, then:
+                //      (f - 1) * 2^e < (w * 10^k) < (f + 1) * 2^e
 
-                    case 23:
-                    case 22:
-                    case 21:
-                    case 20:
-                    {
-                        if (Ten6 <= number)
-                        {
-                            power = Ten6;
-                            exponent = 6;
-                            break;
-                        }
+                DiyFp scaledW = w.Multiply(in tenMk);
 
-                        goto case 19;
-                    }
+                // We now have (double)(scaledW * 10^-mk).
+                //
+                // DigitGenCounted will generate the first requestedDigits of scaledW and return together with a kappa such that:
+                //      scaledW ~= buffer * 10^kappa.
+                //
+                // It will not always be exactly the same since DigitGenCounted only produces a limited number of digits.
 
-                    case 19:
-                    case 18:
-                    case 17:
-                    {
-                        if (Ten5 <= number)
-                        {
-                            power = Ten5;
-                            exponent = 5;
-                            break;
-                        }
+                bool result = DigitGenCounted(in scaledW, requestedDigits, buffer, out length, out int kappa);
+                decimalExponent = -mk + kappa;
+                return result;
+            }
 
-                        goto case 16;
-                    }
+            // Provides a decimal representation of v.
+            // Returns true if it succeeds; otherwise, the result cannot be trusted.
+            //
+            // There will be length digits inside the buffer (not null-terminated).
+            // If the function returns true then:
+            //      v == (double)(buffer * 10^decimalExponent)
+            //
+            // The digits in the buffer are the shortest represenation possible (no 0.09999999999999999 instead of 0.1).
+            // The shorter representation will even be chosen if the longer one would be closer to v.
+            //
+            // The last digit will be closest to the actual v.
+            // That is, even if several digits might correctly yield 'v' when read again, the closest will be computed.
+            private static bool RunShortest(in DiyFp boundaryMinus, in DiyFp w, in DiyFp boundaryPlus, Span<byte> buffer, out int length, out int decimalExponent)
```

Comment:
Based on method description - `TryRunShortest`?

Reply:
Updated all the Grisu3 that return a bool to use `Try` in the name